core/src/main/scala/org/apache/spark/memory/ExecutionMemoryPool.scala

```diff
@@ -138,6 +140,11 @@ private[memory] class ExecutionMemoryPool(
       if (toGrant < numBytes && curMem + toGrant < minMemoryPerTask) {
         logInfo(s"TID $taskAttemptId waiting for at least 1/2N of $poolName pool to be free")
         lock.wait()
+      } else if (toGrant == 0 && memoryFree > 0) {
```

Comment:
This simply means the task has already got assigned the amount of memory that is above average, I don't see why it's still required to wait for more memory here. If your executor memory is not sufficient to support so many tasks, either increase your executor memory or reduce the slots per executor.

Reply:
> I don't see why it's still required to wait for more memory here. If your executor memory is not sufficient to support so many tasks, either increase your executor memory or reduce the slots per executor.

Sometimes required 0 then cause Task throw OOM, in this case always task is heavy, re-compute cost a lot. 

For normal Spark job, we can change config to increase memory or change slot, but for long running Spark such as Thrift server, we can't always restart it.

Reply:
That's exactly what I mean, the executor is oversubscribed by tasks. The task OOM because the executor memory is not enough, the choice here is either to launch worker node with more memory, or change the value of `spark.task.cpus` to allow less tasks run in parallel.

Reply:
Or, if one specific pattern of tasks take significantly more memory than other tasks, maybe you should tune the tasks themselves, instead of modifying the Spark internal for a specific case.

Reply:
> That's exactly what I mean, the executor is oversubscribed by tasks. The task OOM because the executor memory is not enough, the choice here is either to launch worker node with more memory, or change the value of `spark.task.cpus` to allow less tasks run in parallel.

In our case, spark.task.cpus is default value, but core per executor is 4. Active Task not use too much memory decrease the limit of task which use more memory.  So want heavy task wait a little for other task release.  A little corner case

Reply:
> Or, if one specific pattern of tasks take significantly more memory than other tasks, maybe you should tune the tasks themselves, instead of modifying the Spark internal for a specific case.

yea, ad-hoc query sort with big key, spill 20G +. only run that query will success, run with other sql with high concurrency , failed.